core/src/main/java/com/datastax/oss/driver/api/core/paging/Pager.java

```diff
@@ -0,0 +1,324 @@
+import com.datastax.oss.driver.shaded.guava.common.collect.ImmutableList;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+
+public class Pager {
+
+  /** A page returned as the result of an offset query. */
+  public interface Page<ElementT> {
+
+    /** The elements in the page. */
+    @NonNull
+    List<ElementT> getElements();
+
+    /**
+     * The page number.
+     *
+     * <p>It will usually match the requested page number, except if the requested page was past the
+     * end of the iterable, and the out of bounds strategy is {@link
+     * OutOfBoundsStrategy#RETURN_LAST_PAGE}.
+     */
+    int getPageNumber();
+
+    /** Whether there are other pages after this one. */
+    boolean isLast();
+  }
+
+  /** What to do if the caller requests a page that is past the end of the current result set. */
+  public enum OutOfBoundsStrategy {
+    /**
+     * Fail with an {@link IndexOutOfBoundsException}.
+     *
+     * <p>As a special case, if the result set is empty and the first page is requested, this
+     * strategy will return an empty page, not fail.
+     */
+    FAIL,
+
+    /**
+     * Return the last non-empty page of results, instead of the requested page. Note that its
+     * {@link Page#getPageNumber() page number} will be different than the requested one.
+     */
+    RETURN_LAST_PAGE,
+
+    /** Return a page with the correct page number, but no elements. */
+    RETURN_EMPTY_PAGE,
+  }
+
+  private final OutOfBoundsStrategy outOfBoundsStrategy;
+
+  /**
+   * Creates a new instance.
+   *
+   * @param outOfBoundsStrategy the strategy to use if the caller requests a page that is past the
+   *     end of the current result set.
+   */
+  public Pager(@NonNull OutOfBoundsStrategy outOfBoundsStrategy) {
+    this.outOfBoundsStrategy = Objects.requireNonNull(outOfBoundsStrategy);
+  }
+
+  /**
+   * Skips the beginning of a set of results to return the given page, assuming the given page size.
+   *
+   * @param iterable the iterable to extract the results from.
+   * @param targetPageNumber the page to return (1 for the first page, 2 for the second page, etc).
+   *     Must be greater than or equal to 1.
+   * @param pageSize the number of elements per page. Must be greater than or equal to 1.
+   * @throws IllegalArgumentException if the conditions on the arguments are not respected.
+   * @throws IndexOutOfBoundsException if the requested page is past the end of the iterable, and
+   *     the out of bounds strategy is {@link OutOfBoundsStrategy#FAIL}.
+   */
+  @NonNull
+  public <ElementT> Page<ElementT> getPage(
+      @NonNull PagingIterable<ElementT> iterable, final int targetPageNumber, final int pageSize) {
+
+    Objects.requireNonNull(iterable);
+    if (targetPageNumber < 1) {
```

Comment:
In a web UI or REST API it usually starts from 1 (see Google, etc). Since that's a pretty high-level feature we should stay closer to that. See for example how `RandomPagingRestUi` would have to adjust it every time.